impl/src/main/java/org/jboss/weld/xml/BeansXmlParser.java

```diff
@@ -16,144 +16,66 @@
-            if (source.getByteStream().available() == 0) {
-                // The file is just acting as a marker file
-                return EMPTY_BEANS_XML;
-            }
-            BeansXmlHandler handler = getHandler(beansXml);
-
-            try {
-                parser.setProperty("http://java.sun.com/xml/jaxp/properties/schemaLanguage", "http://www.w3.org/2001/XMLSchema");
-                parser.setProperty("http://java.sun.com/xml/jaxp/properties/schemaSource", loadXsds(beansXmlTextBuilder.toString()));
-            } catch (IllegalArgumentException e) {
-                // No op, we just don't validate the XML
-            } catch (SAXNotRecognizedException e) {
-                // No op, we just don't validate the XML
-            } catch (SAXNotSupportedException e) {
-                // No op, we just don't validate the XML
-            }
-
-            parser.parse(source, handler);
-
-            return handler.createBeansXml();
-        } catch (IOException e) {
-            throw XmlLogger.LOG.loadError(beansXml, e);
-        } catch (SAXException e) {
-            throw XmlLogger.LOG.parsingError(beansXml, e);
-        } finally {
-            if (beansXmlInputStream != null) {
-                try {
-                    beansXmlInputStream.close();
-                } catch (IOException e) {
-                    throw new IllegalStateException(e);
-                }
-            }
+        BeansXmlHandler handler = getHandler(beansXml);
+        if (beansXmlValidator != null) {
+            beansXmlValidator.validate(beansXml, handler);
         }
+        return handler != null ? new BeansXmlStreamParser(beansXml, text -> handler.interpolate(text)).parse() : new BeansXmlStreamParser(beansXml).parse();
     }
 
     public BeansXml parse(Iterable<URL> urls) {
         return parse(urls, false);
     }
 
     public BeansXml parse(Iterable<URL> urls, boolean removeDuplicates) {
-        return merge(urls, URL_TO_BEANS_XML_FUNCTION, removeDuplicates);
+        return merge(urls, this::parse, removeDuplicates);
     }
 
-    private <T> BeansXml merge(Iterable<? extends T> items, Function<T, BeansXml> function, boolean removeDuplicates) {
-        List<Metadata<String>> alternatives = new ArrayList<Metadata<String>>();
-        List<Metadata<String>> alternativeStereotypes = new ArrayList<Metadata<String>>();
-        List<Metadata<String>> decorators = new ArrayList<Metadata<String>>();
-        List<Metadata<String>> interceptors = new ArrayList<Metadata<String>>();
-        List<Metadata<Filter>> includes = new ArrayList<Metadata<Filter>>();
-        List<Metadata<Filter>> excludes = new ArrayList<Metadata<Filter>>();
+    protected BeansXmlHandler getHandler(final URL beansXml) {
+        return null;
+    }
+
+    // Merging utils
+
+    public static <T> BeansXml merge(Iterable<? extends T> items, Function<T, BeansXml> function, boolean removeDuplicates) {
```

Comment:
Why do we want to preserve these util methods in a "legacy" class (only kept because of integrators)?
Since you changed the signature, I suppose integrators aren't using it anyway, so we could move it to some util class perhaps?

Reply:
Good point. These should be placed in a separate util class but since they are also used in `BeansXmlParser.parse(Iterable<URL>)` I decided to keep them here.

Reply:
Well, since there is no easy way we can let go of old parser, I suppose there isn't much harm in it.